File: matlab/tests/bench_spectral_ip_gate.m
%% bench_spectral_ip_gate.m
%  Timing-grid harness for the spectral IP gate constant
%  (COST_C in mobius.spectralRelInnerMatrix).
%
%  WHY THIS EXISTS
%  ---------------
%  The gate decides whether the spectral (Fourier) branch of the
%  relative-mode inner product is worth its mode grid against the
%  translation grid's K^2 per event pair. It declines when
%
%      gridSize > COST_C * K^2 * nPairs .
%
%  COST_C absorbs the per-op cost ratio between the two paths, which is
%  a per-language, per-machine quantity: BLAS, JIT, column-major layout
%  and copy-on-write all differ from NumPy, so the constant is fitted on
%  MATLAB timings rather than transcribed from Python.
%
%  This is now purely a SPEED question. Before the full-image work the
%  two routes computed different measures, so a gate that fired on
%  different shapes in the two languages changed the answer; the routes
%  now agree in value (see test_spectral_ip_branch.m), so a mis-set
%  COST_C costs time only. That is why this constant may be recalibrated
%  per machine while the centres/grid gate constants stay matched.
%
%  This harness MEASURES the grid; it does not fit the constant. Run it
%  warm, then send the printed CSV block back so COST_C can be fitted
%  and validated against these exact timings before shipping.
%
%  HOW TO RUN
%  ----------
%  With the toolbox on the path, from anywhere:
%
%      >> bench_spectral_ip_gate
%
%  Run it on an otherwise idle machine. The first call in each cell is
%  discarded to absorb JIT; the reported figure is the median of the
%  rest. Expect a few minutes.
%
%  WHAT IT PRINTS
%  --------------
%  One CSV row per (r, K, sigma/period, isPer) cell:
%
%      r, K, N, sigmaOverP, isPer, gridSize, msSpectral, msGrid, ratio
%
%  ratio = msGrid / msSpectral. The branch is worth taking where
%  ratio > 1. The spectral arm is timed with the cost gate bypassed
%  (internal.spectralIpForce), so every cell measures the branch itself
%  against the grid; the constant is then fitted by routing regret over
%  these timings. A cell whose two arms agree to within noise is one the
%  gate can route either way at no cost.
%
%  This grid stops at K = 30. bench_spectral_ip_gate_ext covers larger K
%  at the shape where the shipped constant was found to misroute.

fprintf('\n=== bench_spectral_ip_gate ===\n');
fprintf('Measuring the spectral branch against the translation grid.\n');
fprintf('Send the CSV block below back for fitting.\n\n');

% The routing announcements ("cos_sim_exp_tens: chose ... path") are
% gated by mptDefaults('showHints'), not by the per-call verbose flag,
% so this is the switch that keeps them out of the CSV. Restored via
% onCleanup so an early exit or Ctrl+C cannot leave the session muted,
% and again explicitly at the end because an onCleanup object created
% by a top-level script survives until its variable is cleared.
bsg_prevShowHints = mptDefaults('showHints', false);
bsg_hintsCleanup  = onCleanup(@() mptDefaults(bsg_prevShowHints)); %#ok<NASGU>

bsg_P      = 1200;
bsg_rs     = [2, 3, 4];
bsg_Ks     = [4, 8, 16, 30];
bsg_Ns     = [1, 2, 4, 8, 16];
bsg_sops   = [0.002, 0.005, 0.0125, 0.05, 0.20];
bsg_isPers = [true, false];
bsg_reps   = 5;

fprintf('CSV_BEGIN\n');
fprintf('r,K,N,sigmaOverP,isPer,gridSize,msSpectral,msGrid,ratio\n');

for bsg_isPer = bsg_isPers
    for bsg_r = bsg_rs
        for bsg_K = bsg_Ks
            if bsg_K < bsg_r
                continue;
            end
            for bsg_N = bsg_Ns
            for bsg_sop = bsg_sops
                bsg_sigma = bsg_sop * bsg_P;

                rng(1000 * bsg_r + 10 * bsg_K + bsg_N);
                % N events per side: the branch's Gram term scales as
                % N_x*N_y while the grid's contraction scales as
                % N_u*N_x*N_y*K^2, so N is the axis the old gate handled
                % worst and the one this sweep exists to cover.
                bsg_p = sort(rand(bsg_K, bsg_N) * bsg_P, 1);
                bsg_q = sort(rand(bsg_K, bsg_N) * bsg_P, 1);
                bsg_w = ones(bsg_K, bsg_N);

                if bsg_isPer
                    bsg_periodArg = bsg_P;
                else
                    bsg_periodArg = 0;
                end

                % Mode-grid size, mirroring the branch's own sizing.
                if bsg_isPer
                    bsg_L = bsg_P;
                else
                    bsg_L = 2 * bsg_P + 2 * (8.6 + 2) * bsg_sigma;
                end
                bsg_M = ceil(8.6 / sqrt(2) * bsg_L ...
                             / (2 * pi * bsg_sigma)) + 2;
                bsg_gridSize = (2 * bsg_M + 1)^(bsg_r - 1);

                % When the mode grid exceeds the memory guard the branch
                % always declines, so both toggle states run the grid and
                % the ratio is 1 by construction. Timing that says nothing
                % about the cost gate and these are the slowest cells, so
                % record the decline without paying for a multi-second
                % grid contraction (matches the Python harness).
                if bsg_gridSize > 4e6
                    fprintf('%d,%d,%d,%.4f,%d,%d,nan,nan,nan\n', ...
                            bsg_r, bsg_K, bsg_N, bsg_sop, bsg_isPer, ...
                            bsg_gridSize);
                    continue;
                end

                % Time the per-attribute inner matrix directly: it is the
                % object the branch replaces, and it takes N events per
                % side without going through a density build.
                bsg_call = @() mobius.relInnerBatched(bsg_p, bsg_w, ...
                    bsg_q, bsg_w, bsg_sigma, bsg_r, bsg_isPer, ...
                    bsg_periodArg, 'truncationSigmas', Inf);

                % Cost-aware repetition: one timed call gauges the cost,
                % then cheap cells get the full bsg_reps for a stable
                % median while cells already costing seconds are timed
                % fewer times (the extended low-sigma/P grid has a few
                % very heavy ones and their magnitude dwarfs jitter).
                % The spectral arm must BYPASS the cost gate, not merely
                % enable the branch: spectralIpEnabled(true) leaves the
                % gate in force, so on every cell the shipped constant
                % declines both arms would run the grid and the ratio
                % would read 1.0 by construction -- which is what an
                % earlier run of this harness reported on all 80 such
                % cells. spectralIpForce bypasses the comparison only,
                % never the MAX_POINTS memory guard.
                internal.spectralIpEnabled(true);
                internal.spectralIpForce(true);
                bsg_t0 = tic; bsg_call(); bsg_first = toc(bsg_t0);
                if bsg_first > 2.0
                    bsg_nrep = 1;
                elseif bsg_first > 0.2
                    bsg_nrep = 2;
                else
                    bsg_nrep = bsg_reps;
                end

                % --- spectral branch ---
                bsg_tS = zeros(1, bsg_nrep);
                for bsg_i = 1:bsg_nrep
                    bsg_t0 = tic;  bsg_call();  bsg_tS(bsg_i) = toc(bsg_t0);
                end

                % --- translation grid ---
                internal.spectralIpForce(false);
                internal.spectralIpEnabled(false);
                bsg_call();
                bsg_tG = zeros(1, bsg_nrep);
                for bsg_i = 1:bsg_nrep
                    bsg_t0 = tic;  bsg_call();  bsg_tG(bsg_i) = toc(bsg_t0);
                end
                internal.spectralIpEnabled(true);

                bsg_msS = 1e3 * median(bsg_tS);
                bsg_msG = 1e3 * median(bsg_tG);
                fprintf('%d,%d,%d,%.4f,%d,%d,%.4f,%.4f,%.4f\n', ...
                        bsg_r, bsg_K, bsg_N, bsg_sop, bsg_isPer, ...
                        bsg_gridSize, bsg_msS, bsg_msG, ...
                        bsg_msG / max(bsg_msS, eps));
            end
            end
        end
    end
end

fprintf('CSV_END\n\n');

mptDefaults(bsg_prevShowHints);
clear bsg_hintsCleanup;
fprintf(['Done. The branch pays off where ratio > 1; COST_C should ' ...
         'place\nthe decline boundary on that contour. Send the block ' ...
         'between\nCSV_BEGIN and CSV_END.\n\n']);

File: python/tools/bench_spectral_ip_gate.py
"""Python twin of matlab/tests/bench_spectral_ip_gate.m.

Times the spectral (Fourier) branch of the relative-mode inner product
against the translation grid, over the same grid of shapes the MATLAB
benchmark uses, and prints a CSV block with identical columns so the two
languages' output can be compared line for line.

It times ``_rel_inner_batched`` directly with the spectral
branch toggled on and off via ``_SPECTRAL_IP_ENABLED`` --- the same
object the MATLAB harness times through ``relInnerBatched`` --- so the
comparison is like for like. Both routes compute the full-image measure,
so the toggle changes cost, not value.

Run from the ``python`` directory of the toolbox, with the package
importable:

    python3 tools/bench_spectral_ip_gate.py

Run it warm on an otherwise idle machine. The first call in each cell is
discarded to absorb any one-off cost; the reported figure is the median
of the rest. Expect a few minutes. Send back the block between
CSV_BEGIN and CSV_END.

Columns: r, K, N, sigmaOverP, isPer, gridSize, msSpectral, msGrid, ratio
where ratio = msGrid / msSpectral (branch worth taking where ratio > 1).
"""
import sys
import time

import numpy as np

from mpt._tensor import cosine as _c
import mpt._tensor._mobius_inner as _mobius_inner

PERIOD = 1200.0
MODE_SIGMAS = _mobius_inner._SPECTRAL_IP_MODE_SIGMAS
REPS = 5

# Same grid as bench_spectral_ip_gate.m.
RS = (2, 3, 4)
KS = (4, 8, 16, 30)
NS = (1, 2, 4, 8, 16)
SOPS = (0.002, 0.005, 0.0125, 0.05, 0.20)
IS_PERS = (True, False)


def _grid_size(sigma, r, is_per):
    # Mirrors the MATLAB harness's reported sizing exactly: the
    # non-periodic embedding length assumes each multiset spans the full
    # period (span_x + span_y = 2 * PERIOD), so the reported gridSize is
    # a fixed function of (sigma, r, is_per) and does not depend on the
    # particular points. This is the sizing bench_spectral_ip_gate.m
    # prints, not necessarily the one the branch computes internally from
    # the actual span; it exists so the two languages' CSV rows carry the
    # same gridSize column and line up.
    if is_per:
        L = PERIOD
    else:
        L = 2.0 * PERIOD + 2.0 * (MODE_SIGMAS + 2.0) * sigma
    M = int(np.ceil(MODE_SIGMAS / np.sqrt(2.0) * L / (2.0 * np.pi * sigma))) + 2
    return (2 * M + 1) ** (r - 1)


def _time(Px, Wx, Py, Wy, sigma, r, is_per, spectral):
    prev = _mobius_inner._SPECTRAL_IP_ENABLED
    prev_force = _mobius_inner._SPECTRAL_IP_FORCE
    _mobius_inner._SPECTRAL_IP_ENABLED = spectral
    # The spectral arm must bypass the cost gate, not merely enable the
    # branch: with the gate in force every cell the shipped constant
    # declines would run the grid in both arms and read a ratio of 1.0
    # by construction (the MATLAB twin was found doing exactly that on
    # all 80 of its declined cells). _SPECTRAL_IP_FORCE bypasses the
    # comparison only, never the _SPECTRAL_IP_MAX_POINTS memory guard.
    _mobius_inner._SPECTRAL_IP_FORCE = bool(spectral)
    try:
        period = PERIOD if is_per else 0.0
        # One warm-up call, discarded, which also gauges the cost.
        t0 = time.perf_counter()
        _mobius_inner._rel_inner_batched(
            Px, Wx, Py, Wy, sigma, r, is_per, period,
            truncation_sigmas=float('inf'))
        first = time.perf_counter() - t0
        # Cost-aware repetition: cheap cells get the full REPS for a
        # stable median; cells already costing seconds are timed fewer
        # times, since their own magnitude dwarfs timing jitter and the
        # extended low-sigma/P grid has a few very heavy ones.
        if first > 2.0:
            reps = 1
        elif first > 0.2:
            reps = 2
        else:
            reps = REPS
        ts = []
        for _ in range(reps):
            t0 = time.perf_counter()
            _mobius_inner._rel_inner_batched(
                Px, Wx, Py, Wy, sigma, r, is_per, period,
                truncation_sigmas=float('inf'))
            ts.append(time.perf_counter() - t0)
        return 1e3 * float(np.median(ts))
    finally:
        _mobius_inner._SPECTRAL_IP_ENABLED = prev
        _mobius_inner._SPECTRAL_IP_FORCE = prev_force


def main():
    sys.stderr.write(
        "\n=== bench_spectral_ip_gate (Python) ===\n"
        "Measuring the spectral branch against the translation grid.\n"
        "Send the CSV block below back for comparison with MATLAB.\n\n")
    sys.stderr.flush()

    print("CSV_BEGIN")
    print("r,K,N,sigmaOverP,isPer,gridSize,msSpectral,msGrid,ratio")
    for is_per in IS_PERS:
        for r in RS:
            for K in KS:
                if K < r:
                    continue
                for N in NS:
                    for sop in SOPS:
                        sigma = sop * PERIOD
                        rng = np.random.default_rng(1000 * r + 10 * K + N)
                        Px = np.sort(rng.uniform(0, PERIOD, (K, N)), axis=0)
                        Py = np.sort(rng.uniform(0, PERIOD, (K, N)), axis=0)
                        Wx = np.ones((K, N))
                        Wy = np.ones((K, N))
                        gs = _grid_size(sigma, r, is_per)
                        # When the mode grid exceeds the memory guard the
                        # branch always declines, so both toggle states run
                        # the grid path and the ratio is 1 by construction.
                        # Timing that adds nothing about the cost gate and
                        # these are the slowest cells, so record the decline
                        # without paying for a multi-second grid contraction.
                        if gs > _mobius_inner._SPECTRAL_IP_MAX_POINTS:
                            print(f"{r},{K},{N},{sop:.4f},{int(is_per)},{gs},"
                                  f"nan,nan,nan")
                            sys.stdout.flush()
                            continue
                        ms_s = _time(Px, Wx, Py, Wy, sigma, r, is_per, True)
                        ms_g = _time(Px, Wx, Py, Wy, sigma, r, is_per, False)
                        ratio = ms_g / max(ms_s, 1e-12)
                        print(f"{r},{K},{N},{sop:.4f},{int(is_per)},{gs},"
                              f"{ms_s:.4f},{ms_g:.4f},{ratio:.4f}")
                        sys.stdout.flush()
    print("CSV_END")
    sys.stderr.write(
        "\nDone. Send the block between CSV_BEGIN and CSV_END.\n\n")


if __name__ == "__main__":
    main()
